### ISSUE TITLE
Create a script for a report that given an author name produces a list of WoS IDs that are returned given the current Sciencewire API and the new WoS API

### ISSUE THREAD

**Issue by peetucket** [sul-dlss/sul_pub#213]
This is useful to determine if the new API is returning comparable data to the current Sciencewire API

For example, pick Russ Altman and Rob Dunbar, run a full dumb query with Sciencewire API, and produce a list of Sciencewire IDs (and hopefully WoSIDs).  Run the same query against the new WoS Client and produce the list of WosIDs.  Ticket is complete when those IDs are available.

**Comment by peetucket** [comment]
Some sample profiles in UAT we can use for comparison of data

- Russ Altman - https://profiles-uat.stanford.edu/russ-altman (586 approved publications)

- Sanjiv Sam Gambhir - https://profiles-uat.stanford.edu/sanjiv-gambhir  (646 approved publications)

- Ann Arvin - https://profiles-uat.stanford.edu/4188?tab=publications (284 approved publications)

- Rob Dunbar - https://profiles-uat.stanford.edu/rob-dunbar?tab=publications (337 approved publications)

- Charbel Farhat - https://profiles-uat.stanford.edu/charbel-farhat?tab=publications (453 approved publications)

- Thomas P. Andriacchi - https://profiles-uat.stanford.edu/thomas-andriacchi?tab=publications (176 approved publications)

**Comment by darrenleeweber** [comment]
WoS API search query for Russ Altman:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:woksearch="http://woksearch.v3.wokmws.thomsonreuters.com" xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
  <soapenv:Body>
    <woksearch:search>
      <queryParameters>
        <databaseId>WOS</databaseId>
        <userQuery>AU=(Altman Russ OR Altman R) AND AD=(Stanford University)</userQuery>
        <timeSpan>
          <begin>1970-01-01</begin>
          <end>2017-09-28</end>
        </timeSpan>
        <queryLanguage>en</queryLanguage>
      </queryParameters>
      <retrieveParameters>
        <firstRecord>1</firstRecord>
        <count>100</count>
        <option>
          <key>RecordIDs</key>
          <value>On</value>
        </option>
        <option>
          <key>targetNamespace</key>
          <value>http://scientific.thomsonreuters.com/schema/wok5.4/public/FullRecord</value>
        </option>
      </retrieveParameters>
    </woksearch:search>
  </soapenv:Body>
</soapenv:Envelope>
```

For that Russ Altman search, the response contains:
```
  <recordsFound>406</recordsFound>
  <recordsSearched>61323448</recordsSearched>
```

**Comment by peetucket** [comment]
```cap_profile_id='12345'
author = Author.find_by_cap_profile_id(cap_profile_id)
author_name = ScienceWire::AuthorName.new(author.last_name,author.first_name,author.middle_name)
institution = ScienceWire::AuthorInstitution.new(Settings.HARVESTER.INSTITUTION.name,ScienceWire::AuthorAddress.new(Settings.HARVESTER.INSTITUTION.address.to_hash))
author_attributes = ScienceWire::AuthorAttributes.new(author_name, '', [], institution)
ids_for_dumb_query = harvester.ids_from_dumb_query(author_attributes)```

**Comment by atz** [comment]
`author` is not really important in that example, except as a place to get `last_name`, `first_name`, and `middle_name` strings.  However it might actually be important because the `harvester` object probably needs to be initialized with the same object.  That is not explicit in the example.

Looking at the difference between "smart" and "dumb" queries, it comes down to:
```ruby
sciencewire_client.query_sciencewire_by_author_name(author_attributes)
# vs.
sciencewire_client.get_sciencewire_id_suggestions(author_attributes)
```
Which in `ScienceWireClient` are, respectively, really just:
```ruby
ScienceWirePublications.new(query_sciencewire(ScienceWire::Query::PublicationQueryByAuthorName.new(author_attributes, max_rows).generate)).filter_publication_items.map(&:publication_item_id)
# and
client.id_suggestions(author_attributes)
```
Too much indirection in this app currently.

Have some preliminary results from console, will post tomorrow.

**Comment by darrenleeweber** [comment]
Work on this report might be able to use some code snippets from https://github.com/sul-dlss/sul_pub/issues/217 to match up the records based on WoS-UID/PMID values.

**Comment by atz** [comment]
### Re: alternate_name_query

I created an Altman fixture record with 2 `author_identities` as in the factory data.

```ruby
[2] pry(main)> author = Author.first
  Author Load (0.2ms)  SELECT  `authors`.* FROM `authors`  ORDER BY `authors`.`id` ASC LIMIT 1
=> #<Author:0x007f97e95733d8
 id: 496,
 cap_profile_id: 98765,
...
 email: "Russ.Altman@stanford.edu",
...
 preferred_first_name: "Russ",
 preferred_last_name: "Altman",
 preferred_middle_name: "Biagio",
...
>
[3] pry(main)> author.author_identities
  AuthorIdentity Load (0.4ms)  SELECT `author_identities`.* FROM `author_identities` WHERE `author_identities`.`author_id` = 496
=> [#<AuthorIdentity:0x007f97e9630f00
  id: 65,
  author_id: 496,
  identity_type: 0,
  first_name: "R",
  middle_name: "B",
  last_name: "Altman",
  email: nil,
  institution: "Stanford University",
...
  created_at: Tue, 03 Oct 2017 20:23:15 UTC +00:00,
  updated_at: Tue, 03 Oct 2017 20:23:15 UTC +00:00>,
 #<AuthorIdentity:0x007f97e9630dc0
  id: 66,
  author_id: 496,
  identity_type: 0,
  first_name: "Russ",
  middle_name: nil,
  last_name: "Altman",
...
  created_at: Tue, 03 Oct 2017 20:23:38 UTC +00:00,
  updated_at: Tue, 03 Oct 2017 20:23:38 UTC +00:00>]
```
Then I did queries.
```ruby
[3] pry(main)> author_name = ScienceWire::AuthorName.new(author.last_name, author.first_name, author.middle_name)
=> #<ScienceWire::AuthorName:0x007f97e95ba0d0 @first="Russ", @last="Altman", @middle="Biagio">
[4] pry(main)> institution = ScienceWire::AuthorInstitution.new(Settings.HARVESTER.INSTITUTION.name, ScienceWire::AuthorAddress.new(Settings.HARVESTER.INSTITUTION.address.to_hash))
=> #<ScienceWire::AuthorInstitution:0x007f97e95e72d8 @address=#<ScienceWire::AuthorAddress:0x007f97e95e73c8 @city="Stanford", @country="USA", @line1="", @line2="", @state="CA">, @name="Stanford University">
[5] pry(main)> author_attributes = ScienceWire::AuthorAttributes.new(author_name, '', [], institution)
=> #<ScienceWire::AuthorAttributes:0x007f97e0d0b860>
[7] pry(main)> harvester = ScienceWire::HarvestBroker.new(author, ScienceWireHarvester.new, alternate_name_query: false)
=> #<ScienceWire::HarvestBroker:0x007f97e1ca2d28 ... > 
[11] pry(main)> ids_for_dumb_query = harvester.ids_from_dumb_query(author_attributes)
=> [53133368, 62190143, 54676777, ... ]
[12] pry(main)> ids_for_dumb_query.count
=> 133
[13] pry(main)> harvester_alt = ScienceWire::HarvestBroker.new(author, ScienceWireHarvester.new, alternate_name_query: true)
=> #<ScienceWire::HarvestBroker:0x007f97e0bfa1b0 ... >
[14] pry(main)> ids_for_dumb_query_with_alternates = harvester_alt.ids_from_dumb_query(author_attributes)
=> [53133368, 62190143, 54676777... ]
[15] pry(main)> ids_for_dumb_query_with_alternates.count
=> 133
[16] pry(main)> ids_for_dumb_query - ids_for_dumb_query_with_alternates
=> []
```

**TLDR**: I get the same 133 records from the old API whether I use `alternate_name_query` or not.  Not sure how to exercise this seemingly important difference in code path.

**Comment by atz** [comment]
Smart query gets the same 133 records:
```ruby
[21] pry(main)> hits = harvester.sciencewire_client.query_sciencewire_by_author_name(author_attributes)
=> [53133368, 62190143, 54676777, ... ]
[22] pry(main)> hits.count
=> 133
[23] pry(main)> ids_for_dumb_query - hits
=> []
```
Same for smart w/ `alternate_name_query` in this case.

**Comment by atz** [comment]
Note: for the purpose of this ticket, the starting point for comparison in the old API will be "dumb" query WITH `alternate_name_query`.

(Even though the results are the same on this one case, they may diverge in others.)

**Comment by darrenleeweber** [comment]
```ruby
wos_client = WosClient.new(Settings.WOS.AUTH_CODE, :info);
wos_queries = WosQueries.new(wos_client);
records = wos_queries.search_by_name('Altman, Russ');
puts records.uids.to_s
["MEDLINE:24551397", "WOS:000172527000005", "MEDLINE:12647379", "MEDLINE:20161212", "MEDLINE:2611273", "MEDLINE:18819074", "WOS:000251034700001", "WOS:000306925000001", "WOS:000290218300018", "WOS:000240964500001", "WOS:000273866500001", "WOS:000259179300003", "WOS:000337000402635", "WOS:000222244300001", "WOS:000236528700024", "MEDLINE:8130522", "WOS:000245271200003", "WOS:000183123802264", "WOS:000224023200028", "MEDLINE:26776202", "WOS:000235327300007", "WOS:000279559500695", "WOS:000173252700826", "WOS:000083035800004", "WOS:000399659000008", "WOS:000249636500024", "WOS:000242874200010", "WOS:A1993LP80900004", "WOS:000076973800010", "WOS:A1996TW69600005", "WOS:000179509700007", "WOS:000227190000011", "WOS:000321153600012", "WOS:000384536000001", "WOS:000281003900006", "WOS:000289592600005", "WOS:000365709900005", "WOS:000336507500015", "WOS:000176553400006", "WOS:000358821300020", "WOS:000272464000008", "WOS:000299167800043", "WOS:000257860800004", "WOS:000250712600021", "WOS:000273849100001", "WOS:000403782300001", "WOS:000183199900007", "WOS:000322064400019", "WOS:000348562700003", "MEDLINE:24452614", "WOS:000087898000028", "WOS:000336415300030", "MEDLINE:25717414", "WOS:000386326200035", "WOS:000260325500002", "MEDLINE:24303232", "WOS:000265602500007", "WOS:000335405200021", "WOS:000284199500014", "WOS:000296146400010", "WOS:000272064000111", "WOS:000312901500023", "WOS:000346632900006", "WOS:A1995RN00200014", "MEDLINE:24297550", "WOS:000175155500002", "WOS:000276769300010", "WOS:000271425300004", "WOS:000300768100009", "WOS:000169404700005", "WOS:000285036700017", "WOS:000328629800009", "WOS:000312618200018", "WOS:000304245800017", "WOS:000253033000002", "WOS:000309017000009", "MEDLINE:7949944", "WOS:A1994BA84G00018", "WOS:000220143500001", "WOS:A1989CW82100011", "WOS:A1995TR87100009", "WOS:000267225200003", "WOS:000224359100003", "MEDLINE:7584378", "MEDLINE:9783220", "WOS:A1994BC13B00068", "MEDLINE:7584390", "WOS:000287439600011", "MEDLINE:7584327", "WOS:000368357800013", "WOS:A1990EF71200016", "WOS:000072320000002", "MEDLINE:10977090", "WOS:000165970200020", "WOS:000227313200021", "WOS:000364401600008", "WOS:000258773600005", "WOS:000298848100012", "WOS:000328951000007", "WOS:000336517800016", "WOS:000263639700041", "WOS:000172328100058", "WOS:000290218700009", "WOS:000301538300005", "WOS:000304437400001", "WOS:000246024700031", "WOS:000298488200009", "WOS:000075921100005", "WOS:000251834500007", "WOS:000316833900008", "WOS:000284148300006", "WOS:000309977100008", "WOS:000291752600050", "WOS:000285036700009", "WOS:000268214500004", "WOS:000326480000024", "WOS:000281927200010", "WOS:000179249800008", "WOS:000244446000202", "WOS:000407150800040", "WOS:000292061700004", "WOS:000278125200026", "WOS:000254263300015", "WOS:000360620100003", "WOS:000184532900029", "WOS:000276506900009", "WOS:A1988R230100006", "WOS:000397264100008", "WOS:000256421800001", "WOS:000222501000030", "WOS:A1995QU44000004", "WOS:000271602800010", "WOS:000352195700026", "WOS:000286192100001", "WOS:000225497400003", "WOS:000360322800016", "WOS:000293691400076", "WOS:000254800400002", "WOS:000247465400004", "WOS:000180913600011", "WOS:000226723300159", "WOS:000243893500009", "WOS:000276769300008", "WOS:000183970000023", "WOS:000184532900040", "WOS:000183832900010", "WOS:000335765500022", "WOS:000348126800006", "WOS:000293620800028", "WOS:000262463200001", "WOS:A1986F079500001", "WOS:A1995RL45300002", "WOS:000072320000011", "WOS:000082447300006", "WOS:A1991GP84100006", "WOS:A1996VM02500008", "WOS:000089431900002", "WOS:A1990DN23200002", "WOS:A1992JF96900006", "WOS:000225031500008", "WOS:000260433100007", "WOS:000239308600017", "WOS:000252545400160", "MEDLINE:16100408", "WOS:000256961700006", "WOS:000207722807509", "WOS:000277890200036", "WOS:000266575500001", "WOS:000312901500006", "WOS:000189418100092", "WOS:A1991FJ10000002", "WOS:A1997XC85400014", "WOS:000171768600009", "WOS:000181756700019", "WOS:A1996WC40600007", "WOS:000231800100001", "WOS:A1994QF21600004", "WOS:A1997WR20700003", "WOS:A1995RT03000003", "WOS:000171768600146", "MEDLINE:28412959", "WOS:000224303500010", "WOS:000292634200009", "MEDLINE:19908383", "WOS:000359645700006", "WOS:000281830900010", "WOS:000267619000010", "MEDLINE:18229697", "WOS:000407150800033", "MEDLINE:25592562", "WOS:000289460200009", "MEDLINE:12463865", "WOS:000352371900019", "WOS:000286971900007", "WOS:000176159200048", "WOS:000293731200012", "WOS:000285331700007", "MEDLINE:20351936", "WOS:000295896300011", "WOS:000274306700011", "WOS:000267619000009", "WOS:000171950200005", "MEDLINE:7949915", "WOS:A1997WD20100027", "WOS:000303769700007", "WOS:000386034800127", "MEDLINE:23824865", "WOS:000187587700002", "WOS:000308887200005", "WOS:000253779800016", "WOS:000298249500009", "WOS:000243216000007", "WOS:000184222500057", "WOS:000316833300001", "WOS:000304924100001", "WOS:000302783800008", "WOS:000303930500003", "WOS:000181303000011", "WOS:000355286600015", "WOS:000306483500009", "WOS:000278879400009", "MEDLINE:19904832", "MEDLINE:19908374", "WOS:000173064900022", "MEDLINE:10902193", "WOS:000374412300003", "WOS:000375514000018", "MEDLINE:12603029", "WOS:000348730500184", "WOS:A1994QF21600059", "MEDLINE:9929182", "MEDLINE:8947767", "MEDLINE:7584427", "MEDLINE:7950031", "MEDLINE:9929358", "MEDLINE:7949909", "WOS:000274097500055", "WOS:A1989AU61400004", "WOS:000170207500089", "WOS:A1994QF21600137", "WOS:000235019400028", "MEDLINE:8947710", "MEDLINE:9697207", "WOS:A1996BH75M00047", "MEDLINE:9390315", "WOS:A1994QF21600033", "MEDLINE:10902175", "MEDLINE:9929319", "WOS:000275410900001", "WOS:000386326200019", "WOS:000317477500001", "WOS:000189389100013", "WOS:000168716800008", "WOS:000356780900020", "WOS:000291633300011", "WOS:000307652600006", "WOS:000381691000001", "WOS:000186175900001", "WOS:000277655100025", "WOS:000355751600011", "WOS:000357993500007", "WOS:000405616900002", "WOS:000389557000012", "WOS:000291853800023", "WOS:000383781600015", "WOS:000359895400001", "MEDLINE:12831574", "WOS:000408790100013", "MEDLINE:24516403", "WOS:000331209100006", "WOS:000185086100017", "WOS:000080967100007", "WOS:000300409800008", "MEDLINE:28458783", "WOS:000399806500016", "WOS:000288444500010", "WOS:000301537400010", "WOS:000337000400071", "WOS:000409183600006", "WOS:000336260000006", "WOS:000334429300007", "WOS:000279865400007", "MEDLINE:22867428", "WOS:000356370900005", "WOS:000323226500009", "WOS:000185701100022", "MEDLINE:21121048", "WOS:000324133500001", "WOS:000325690200011", "WOS:000330879300006", "WOS:000323220200007", "WOS:000268565100019", "WOS:000171768600185", "WOS:000081953300015", "WOS:000266575500002", "WOS:000245882400002", "WOS:000278879400001", "WOS:000257188700033", "WOS:000290431200007", "WOS:000332021500002", "WOS:000180584000002", "WOS:A1996VH69500001", "WOS:000376584400019", "WOS:000178396400014", "WOS:000221171700016", "WOS:000332575800021", "WOS:000247557800010", "WOS:000277594800007", "WOS:000079090200006", "WOS:000275061200007", "WOS:000286096000006", "MEDLINE:11262956", "WOS:000297621501476", "WOS:000281388500002", "MEDLINE:9322019", "WOS:000392723700004", "WOS:000405361900008", "WOS:000255026100040", "WOS:000347393200006", "WOS:000295419100031", "WOS:000301889500023", "WOS:000400664400006", "WOS:000405498300005", "WOS:000311031800008", "WOS:000359428300001", "WOS:000285410500012", "WOS:000281295500008", "WOS:000273307600008", "WOS:000282965100006", "WOS:000267619000011", "WOS:000276373800008", "WOS:000267619000007", "WOS:000182161400002", "WOS:000085267900007", "WOS:000317477500029", "WOS:000254012100001", "WOS:000305429900009", "WOS:000258424100003", "WOS:000296799900016", "WOS:000324527600007", "MEDLINE:11928517", "WOS:000325690200010", "WOS:000326971400009", "WOS:000366135900002", "MEDLINE:26306281", "WOS:000328629800008", "WOS:000337727000007", "WOS:000343666300003", "WOS:000082944500017", "WOS:000209001600333", "MEDLINE:27199190", "WOS:000225769500002", "WOS:000227842000009", "WOS:000276704800009", "WOS:000237567000021", "WOS:000178914400005", "WOS:000267619000008", "WOS:000319869500002", "WOS:000233691100005", "WOS:000178338600006", "WOS:000227842000003", "WOS:000340731400006", "WOS:000394450500004", "WOS:000246464800017", "WOS:000373526700005", "WOS:000316109700008", "WOS:000387469100004", "WOS:000398197800006", "WOS:000349450800007", "WOS:000355557900008", "WOS:000342763900023", "WOS:000351668600006", "WOS:000364626100006", "WOS:000390851300005", "WOS:000380831700005", "WOS:000299310600008", "WOS:000309017000017", "WOS:000233119600015", "WOS:000279019500008", "WOS:000319869500011", "WOS:000407150800044", "WOS:000239480500002", "MEDLINE:26776186", "WOS:000339166700007", "WOS:000309146200001", "WOS:000188389700012", "WOS:000272310800008", "WOS:000319869500006", "WOS:000292681800008", "WOS:000186080501585", "WOS:000303223300594", "WOS:000339602900018", "WOS:A1991FF99400004", "WOS:000309115000007", "WOS:000398829200005", "WOS:000294808900008", "WOS:000262268500012", "WOS:000353198700012", "WOS:000359752100030", "WOS:000258751400003", "WOS:A1994NF04400002", "WOS:000336509000034", "WOS:000289984100006", "WOS:000234555300014", "WOS:000292843000016", "WOS:000332858400048", "WOS:000315452000017", "WOS:000358581700030", "WOS:000224551401078", "WOS:000325359205087", "WOS:000238521700009", "WOS:000187515400027", "WOS:000178421800365", "WOS:000344384900244", "WOS:000185432801177", "WOS:000178421800344", "WOS:000344384900249", "WOS:000314138700072", "WOS:000224551500220", "WOS:000245081900001", "WOS:000278173900007", "WOS:000176277100001", "WOS:000238946500027", "WOS:000295119200035", "WOS:000238959700012", "WOS:000177087200015", "WOS:000324068800003", "WOS:000385659500007", "WOS:000223799000910", "WOS:000353821200005", "WOS:000358434600025", "WOS:000242387100004", "WOS:000221546700001", "WOS:000362838700024", "WOS:000085609800009", "WOS:000183091100006", "WOS:000222904300001", "WOS:000071909500020", "WOS:A1997WR17000036", "WOS:000165771000001", "WOS:000295770300001", "WOS:000237917900001", "WOS:000268735900001", "WOS:A1987L238800003", "MEDLINE:24516805", "WOS:000188978800018", "WOS:000367655600002", "WOS:000226524300059", "WOS:000277335900027", "WOS:000334741600026", "WOS:000324239200029", "WOS:A1987H850000003", "WOS:A1986D612900016", "WOS:000227950600009", "WOS:000335874800038", "WOS:A1990EJ79900001", "WOS:000330151100026", "WOS:000311283400016", "WOS:000236700200001", "WOS:000322064400018", "WOS:000407082400024", "WOS:000263411300005", "WOS:000244850300011", "WOS:000326656200047"]
```

**Comment by atz** [comment]
This ticket shouldn't have been closed.  The title is: `Produce a report that given an author name produces...`  Was a report produced?  No.

**Comment by peetucket** [comment]
Note, I accepted this  - the idea being the "report" was the list of IDs, and I created a new ticket to do some analysis: https://github.com/sul-dlss/sul_pub/issues/261
We can discuss more at standup about how to proceed

**Comment by atz** [comment]
@peetucket why is matching via the `PublicationIdentifiers` table a valid way to measure the intersection of two different semantic queries?  AFAIK, we have no guarantee the values there were written as the result of any given query (e.g. batch vs. manually added, or the result of one form of the author's name vs. another, or vs. several combined).  `PublicationIdentifiers` can tell us "the new query returns stuff we recognize", but not (by itself) what the relationship to other IDs currently might be.  AFAICT, we have to go **through** that table to get the sciencewire IDs, like:
```ruby
PublicationIdentifier.where(...).map { |x| x.publication.sciencewire_id }
```

Then it's possible to compare.

FYI: production DB currently has these counts for `PublicationIdentifiers#identifier_type`:
```ruby
"doi" => 194556, 
"isbn" => 2456, 
"legacy_cap_pub_id" => 7751, 
"pmc" => 58352, 
"PMID" => 260173, 
"PublicationItemID" => 377025, 
"SULPubId" => 8114, 
"WoSItemID" => 333002
```

**Comment by atz** [comment]
@peetucket: The ONLY place I've found where the `publication.sciencewire_id` matches the `publication_identifiers.identifier_value` is when `identifier_type = 'PublicationItemID'`.  So for the two queries with `identifier_type: 'WoSItemID'` and `identifier_type: 'PMID'`, why would presence in the table imply that a record has the SAME ScienceWire ID?  Or ANY?

The assumptions seem to be:
- A matching PMID or WosItemID in `publication_identifiers` means the record MUST have a ScienceWire ID.
- That ScienceWire ID MUST match whatever the ScienceWire API is currently returning.

Those are demonstrably **false**.  Example, `.where("publications.sciencewire_id IS NULL").group(:identifier_type).count`:
```ruby
{"doi"=>10166, "isbn"=>2456, "legacy_cap_pub_id"=>7749, "pmc"=>822, "PMID"=>4725, "SULPubId"=>8114}
```
So the good news is `WoSItemID`s all have ScienceWire IDs and `PublicationItemID`s ARE ScienceWire IDs, but the bad news is that every other "type" has records without ScienceWire IDs, including PubMed.
